### PR TITLE
Add functionality to cols for incoming data

### PR DESCRIPTION
**Pull request by gcoffmanQ**
We encountered a problem in prod where new data that is being ingested via the CDP Refresh job has different column types than is in the entity table: https://usermind-prod.cloud.databricks.com/?o=661009023962603#job/69/run/11087590. For some reason, the data is now String's in the data but is Double's in the entity table. This PR is to fix this by automatically casting columns for incoming data that don't match the column type in the entity table.

We want to be really careful when doing this casting, so we first check how many nulls are in that column before the cast, perform the cast, then check if there is a different number of nulls. If the null counts check, only then do we return the result of the cast.

### CODE REVIEW

File: cdp-refresh.py [hunk 1]
```diff
@@ -95,8 +102,63 @@ def fix_spark_sql_missing_columns(org_id, entity_table):
...
+    existing_null_count = get_null_count(col_name, table_col_type, null_count_query)
+    cast_null_count = get_null_count(col_name, table_col_type, null_count_cast_query)
+    print(f"Existing null count in col {col_name} in new data was: {existing_null_count} and null count after casting type to {table_col_type}: {cast_null_count}")
+    return existing_null_count == cast_null_count
+    
+def get_null_count(col_name, table_col_type, query):
+    print(f"{col_name}, {table_col_type}, Null count query: {query}")
+    null_rows = spark.sql(query).collect()
+    if len(null_rows) == 0:
+        return 0
```
Comment (aruni-xm) [inline cdp-refresh.py:157]: if we fix the query to have count, this branch should never be taken

File: cdp-refresh.py [hunk 2]
```diff
@@ -28,6 +32,9 @@
 jdbc_database = 'insights'
 jdbc_port = '5432'
 
+NULL_COUNT_QUERY = "SELECT COUNT({col_name}) FROM new_data WHERE {col_name} = null"
+CAST_NULL_COUNT_QUERY = "SELECT * FROM (SELECT CAST({col_name} AS {table_col_type}) FROM new_data) WHERE {col_name} = null"
```
Comment (aruni-xm) [inline cdp-refresh.py:36]: these lines are repeated.

**Review by cfluture — APPROVED** [review]

File: cdp-refresh.py [hunk 3]
```diff
@@ -95,8 +102,63 @@ def fix_spark_sql_missing_columns(org_id, entity_table):
...
+def construct_col_type_map(schema):
+    col_dict = {}
+    for col in schema:
+        col_dict[col.name] = col.dataType
+    return col_dict
+
+def change_data_types(org_id, entity_table, new_data_df, col_type_diff_dict):
+    for col_name, table_col_type in col_type_diff_dict.items():
+        print(f"Trying to convert col {col_name} to table col type of: {table_col_type}")
+        table_col_type = str(table_col_type).split("Type")[0]
```
Comment (aruni-xm) [inline cdp-refresh.py:131]: nit: please extract this in a function (can come in a follow-up PR). Also does this work for all Spark types?

File: cdp-refresh.py [hunk 4]
```diff
@@ -140,3 +204,19 @@ def fix_spark_sql_missing_columns(org_id, entity_table):
...
+from pyspark.sql.types import *
+
+df = spark.table("org_1.sftp_mktodata_1055269_backup")
+org_id = "org_1"
+table_name = "sftp_mktodata_1055269_backup"
+new_df = df.withColumn("EventId", F.col("EventId").cast(DoubleType()))
+#new_df.show()
+fixed_df = fix_new_data_col_types(org_id, table_name, new_df)
+fixed_df.show()
+
```
Comment (aruni-xm) [inline cdp-refresh.py:219]: remove test code

**Review by aruni-xm — APPROVED** [review]

**Pull request merged**